docs: add debian installation guide

CDLuminate opened the pull request:
This PR closes issue: https://github.com/bvlc/caffe/issues/2601

williford commented:
@CDLuminate I made a few changes in my submit. I hope they are welcomed. I haven't tried the instructions, but it looks good to me.

CDLuminate commented on docs/install_apt_debian.md:
```diff
@@ -4,10 +4,11 @@ title: "Installation: Debian"
 
-Caffe packages are available for `Debian/unstable`. Debian/stable users
-should take a look at Ubuntu installation instruction.  
+Caffe packages are available for `Debian/unstable` (Debian 10). Debian/stable
```
I think there is no need to refer `Debian/unstable` such a rolling release as `Debian 10`. And actually Debian/unstable is currently Stretch (Debian 9).

williford replied:
Stretch is `Debian/testing` (https://wiki.debian.org/DebianTesting & https://wiki.debian.org/DebianReleases#Current_Releases.2FRepositories), no? Debian/unstable will have the codename "buster", after it gets promoted from being unstable/sid. If the codenames are given later on, I think it is fine to leave it off.

CDLuminate commented on docs/install_apt_debian.md:
```diff
@@ -4,10 +4,11 @@ title: "Installation: Debian"
-Caffe packages are available for `Debian/unstable`. Debian/stable users
-should take a look at Ubuntu installation instruction.  
+Caffe packages are available for `Debian/unstable` (Debian 10). Debian/stable
+(jessie) and Debian/testing (stretch) users should take a look at [Ubuntu
```
Currently Debian/testing users should be able to install these caffe packages. With a sources.list like this and a configured `apt_preferences`(man apt_preferences) :
```
deb http://mirror/debian testing main contrib non-free
deb http://mirror/debian unstable main contrib non-free
```
I didn't refer this because this is advanced usage, and it may break after several months.

There is still hope that Caffe can enter Debian/Stretch, if some warm-hearted volunteer can fix the skimage package.

https://tracker.debian.org/pkg/caffe

williford replied:
Okay, I was confused about it being left off, so I just guessed :-).

williford replied:
@CDLuminate I thought that I would be able to edit or revert my commit to your PR, but it looks like I can only make another commit changing my previous one. I think it would be good to at least mention that by "testing", you mean "stretch", but the other changes are not necessary. Would you mind reverting my commit? Sorry for the hassle.

CDLuminate commented:
@williford Thank you for the change. It looks good to me.

CDLuminate commented:
Well it seems that several Debian developers are fixing the skimage package in the last minutes.

CDLuminate commented:
skimage package was fixed by a Debian science developer. So I expect to see `caffe-cpu` in Debian/Stretch.

@shelhamer @lukeyeager  Please review this document when you are available. I'll update this document when I finally confirmed the result of Debian/Stretch (testing).

lukeyeager commented on docs/install_apt_debian.md:
```diff
@@ -0,0 +1,108 @@
+
+Only experienced linux users are recommended to try Debian/unstable (Sid). 
+
+Last update: Dec.21 2016  
```
Odd date format. How about 2016-12-21?
https://xkcd.com/1179/

lukeyeager commented on docs/install_apt_debian.md:
```diff
@@ -0,0 +1,108 @@
+
+`sudo apt install caffe-cuda`, apt's dependency resolver is smart enough to deal with this.
+
+* Where is the examples, the models and other documentation stuff?
```
How about "Where **are** the examples, models and other documentation?"

lukeyeager commented on docs/installation.md:
```diff
@@ -12,6 +12,7 @@ The official Makefile and `Makefile.config` build are complemented by a [communi
 
 - [Docker setup](https://github.com/BVLC/caffe/tree/master/docker) *out-of-the-box brewing*
 - [Ubuntu installation](install_apt.html) *the standard platform*
+- [Debian installation](install_apt_debian.html) *deploy caffe with a single command*
```
How about "install" instead of "deploy"?

CDLuminate commented:
@lukeyeager Thank you for the comments. I've updated the guide according to your suggestions.

shelhamer commented:
@CDLuminate looks good to me, and I expect this should be helpful to the debian part of the community. Could you just squash the @williford commits into one to clean up the history?

CDLuminate commented:
@shelhamer I've squashed @williford commits into one, and added two tables to the doc.

One indicating if they are able to do APT install, another indicating if their compiler combination would work.

```
Your Distro     |  CPU_ONLY  |  CUDA  |     Alias
----------------+------------+--------+-------------------
Debian/stable   |     ☒      |   ☒    | Debian Jessie
Debian/testing  |     ☑      |   ☐    | Debian Stretch/Sid
Debian/unstable |     ☑      |   ☑    | Debian Sid
```

```
CXX compiler |  CUDA 7.5  |  CUDA 8.0  |
-------------+------------+------------+-
GCC-7        |     ?      |     ?      |
GCC-6        |     ☒      |     ☒      |
GCC-5        |     ☑ [1]  |     ☑      |
CLANG-4.0    |     ?      |     ?      |
CLANG-3.9    |     ☒      |     ☒      |
CLANG-3.8    |     ?      |     ☑      |
```

CDLuminate commented:
Oh wait, I can't see \u0x2611 ☑ and \u2612 ☒ clearly on github with the default scaling factor.

CDLuminate commented:
Better after `%s/☑/✔/g` and `%s/☒/✘/g`

CDLuminate commented:
non-official version: https://cdluminate.github.io/caffe.html

shelhamer commented:
Thanks for all of the debian packaging work @CDLuminate !

CDLuminate commented:
LOL

CDLuminate commented:
I checked the git history. Currently Debian's Caffe version `1.0.0~rc3+20161127-g24d2f67-*` is nearly identical to [1.0.0~rc4](https://github.com/BVLC/caffe/releases/tag/1.0.0-rc4) the silent release, except for some typo fixes.

This version (I mean `1.0.0~rc3+20161127-g24d2f67` where "g24d2f67" means git24d2f67) will keep staying in Debian Stretch untill its end of life (Several years). However the Caffe version in Debian Sid will keep rolling with you.

Now let's talk about Ubuntu. Theoretically speaking Ubuntu keeps pulling most of the pacckages directly from Debian unstable to its devel branch, then freeze, and release when it's time. Now my package simply doesn't build on Ubuntu devel and I still don't know the reason why it failes in the unit test. [1]

[1] This is an amd64 build failure example. Failed to build from source because a unit test failure. https://launchpadlibrarian.net/302187072/buildlog_ubuntu-zesty-amd64.caffe_1.0.0~rc3+20161127-g24d2f67-5_BUILDING.txt.gz

@shelhamer

cypof commented:
The rc4 tag was just to have one ready now for the docker [update](https://github.com/BVLC/caffe/pull/5153), but it was meant to include the NCCL work too, which should merge pretty soon. I will move it afterward and then we can post something about it.

shelhamer commented:
@CDLuminate @cypof rc4 isn't supposed to exist yet—I'm not sure how that was made. I hate to switch a tag once it exists but that was in error, and we'll define it once with the usual release notes once NCCL is merged (real soon now).

CDLuminate commented:
@shelhamer If the real rc4 release can be made before Jan 24, it will be possible to push rc4 into stretch.

shelhamer commented:
@CDLuminate rc4 is out: https://github.com/BVLC/caffe/releases/tag/rc4

CDLuminate commented:
I updated caffe-cpu to rc4 version. Will update caffe-cuda tomorrow.

caffe-cpu_rc4 is available for amd64, armel, i386, ppc64el, s390x now.

https://buildd.debian.org/status/package.php?p=caffe
